core/geometry/src/polyface/PolyfaceBuilder.ts

```diff
@@ -571,55 +573,55 @@ export class PolyfaceBuilder extends NullGeometryHandler {
       if (vectorAC.magnitude() >= vectorBD.magnitude()) {
         this.addTriangleFacet(
           [points[0], points[1], points[2]],
-          needParams ? [param0!, param1!, param2!] : undefined,
-          needNormals ? [normal0!, normal1!, normal2!] : undefined,
-          needColors ? [color0!, color1!, color2!] : undefined,
+          needParams ? [param0, param1, param2] : undefined,
+          needNormals ? [normal0, normal1, normal2] : undefined,
+          needColors ? [color0, color1, color2] : undefined,
```

Comment:
If `needsColors` is true and `colors` is `undefined` or has less than 4 values, you wind up with the `0` color values here. I suspect that an else clause for the if statement on line 561 should be added that throws an exception.